flink-runtime/src/test/java/org/apache/flink/runtime/taskexecutor/TaskExecutorPartitionLifecycleTest.java

```diff
@@ -0,0 +1,279 @@
+			})
+			.build();
+
+		final TestingTaskExecutor taskManager = createTestingTaskExecutor(taskManagerServices, new HeartbeatServices(Long.MAX_VALUE, Long.MAX_VALUE));
+
+		try {
+			taskManager.start();
+			taskManager.waitUntilStarted();
+
+			final String jobMasterAddress = "jm";
+			rpc.registerGateway(jobMasterAddress, jobMasterGateway);
+
+			// inform the task manager about the job leader
+			jobLeaderService.addJob(jobId, jobMasterAddress);
+			jobManagerLeaderRetriever.notifyListener(jobMasterAddress, UUID.randomUUID());
+
+			while (!taskManagerServices.getJobManagerTable().contains(jobId)) {
+				Thread.sleep(50);
+			}
+
+			final ExecutionAttemptID eid1 = new ExecutionAttemptID();
+			final IntermediateResultPartitionID partitionId = new IntermediateResultPartitionID();
+			final ResultPartitionID resultPartitionId = new ResultPartitionID(partitionId, eid1);
+			final ResultPartitionDeploymentDescriptor task1ResultPartitionDescriptor =
+				new ResultPartitionDeploymentDescriptor(
+					new PartitionDescriptor(new IntermediateDataSetID(), partitionId, ResultPartitionType.BLOCKING, 1, 0),
+					new ShuffleDescriptor() {
+						@Override
+						public ResultPartitionID getResultPartitionID() {
+							return resultPartitionId;
+						}
+					},
+					1,
+					true);
+
+			final TaskDeploymentDescriptor taskDeploymentDescriptor =
+				TaskExecutorSubmissionTest.createTaskDeploymentDescriptor(
+					jobId,
+					"job",
+					eid1,
+					new SerializedValue<>(new ExecutionConfig()),
+					"Sender",
+					1,
+					0,
+					1,
+					0,
+					new Configuration(),
+					new Configuration(),
+					TestInvokable.class.getName(),
+					Collections.singletonList(task1ResultPartitionDescriptor),
+					Collections.emptyList(),
+					Collections.emptyList(),
+					Collections.emptyList(),
+					0);
+
+			taskSlotTable.allocateSlot(0, jobId, taskDeploymentDescriptor.getAllocationId(), Time.seconds(60));
+
+			TestInvokable.sync = new BlockerSync();
+
+			taskManager.submitTask(taskDeploymentDescriptor, jobMasterGateway.getFencingToken(), timeout)
+				.get();
+
+			TestInvokable.sync.awaitBlocker();
+
+			// the task is still running, so nothing was actually stored yet
+			assertTrue(jobAwareShuffleEnvironment.hasPartitionsOccupyingLocalResources(jobId));
+
+			TestInvokable.sync.releaseBlocker();
+			taskFinishedFuture.get(timeout.getSize(), timeout.getUnit());
+
+			// the task is still running, so nothing was actually stored yet
+			assertTrue(jobAwareShuffleEnvironment.hasPartitionsOccupyingLocalResources(jobId));
+
+			releaseAction.accept(jobId, new ResultPartitionID(partitionId, eid1), taskManager);
+			postReleaseActionAssertion.accept(jobId, jobAwareShuffleEnvironment);
+		} finally {
+			RpcUtils.terminateRpcEndpoint(taskManager, timeout);
+		}
+
+		// the partition table should always be cleaned up on shutdown
+		assertFalse(jobAwareShuffleEnvironment.hasPartitionsOccupyingLocalResources(jobId));
```

Comment:
put this in finally after `RpcUtils.terminateRpcEndpoint`?

Reply:
I'd keep it here; if the try block fails with an exception we don't want additional assertions obfuscating the actual problem.

Reply:
I thought even though an exception occurs and as long as the finally shutdown is executed, this assert should be right, so might not bring noise.

Sure, you could keep the current way and I have no preference. :)